### PR TITLE
adding first draft of doctr integration

### PR DESCRIPTION
**Pull request by ryanchesler**
adding doctr as an option for OCRing images

Our doctr branch has access to more accurate models and faster inference on GPUs as well as onnx/openvino fast inference on CPU. Using our heaviest and most accurate model, but can be configured to allow using much smaller and faster variants

Used a similar pattern to the captions model

### CODE REVIEW

**Comment by ryanchesler** [comment]
found an issue in run_qa_db. Will have to spend some more time on it

**Comment by ryanchesler** [comment]
resolved issues related to this, but found another issue. Seems that the options selected dont propagate if you upload a zip. It always reverts to the default blip.

**Comment by ryanchesler** [comment]
added an additional optional requirements file for doctr that needs to be installed beforehand

File: reqs_optional/requirements_optional_doctr.txt [hunk 1]
```diff
@@ -0,0 +1 @@
+python-doctr @ git+https://github.com/h2oai/doctr.git@9d0b160ee80b38d8c12672bde78aed7965054567
```
Comment (pseudotensor) [inline reqs_optional/requirements_optional_doctr.txt:1]: maybe just FYI.  When I install this, pip complained only about argilla-1.12.0, then I upgrade that to 1.14.1 then it installs pandas 1.5.3.  I presume we want pandas 2.0, so probably some issue check on.

On a different computer, instead if complained about:
```
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
apache-beam 2.46.0 requires dill<0.3.2,>=0.3.1.1, but you have dill 0.3.6 which is incompatible.
```

**Comment by pseudotensor** [comment]
Running tests overnight with this PR on main, to at least see if the doctr import is heavy and messes with things.  Do you know if the doctr import loads-up cuda or other packages or even starts-up some models that would break multiprocessing or asyncio etc.?

We can move import to deeper if issues.

**Comment by pseudotensor** [comment]
Image:

driverslicense


Tesseract:

Fails with DocTR installed.

```

Failed to ingest /tmp/gradio/9d1c1cdef3600e9ab0687adc5b67001c8ecbabbd/driverslicense.jpeg due to Traceback (most recent call last):
  File "/home/jon/h2ogpt/src/gpt_langchain.py", line 1944, in path_to_doc1
    res = file_to_doc(file, base_path=None, verbose=verbose, fail_any_exception=fail_any_exception,
  File "/home/jon/h2ogpt/src/gpt_langchain.py", line 1635, in file_to_doc
    docs1a = UnstructuredImageLoader(file, strategy='hi_res').load()
  File "/home/jon/miniconda3/envs/h2ollm/lib/python3.10/site-packages/langchain/document_loaders/unstructured.py", line 86, in load
    elements = self._get_elements()
  File "/home/jon/miniconda3/envs/h2ollm/lib/python3.10/site-packages/langchain/document_loaders/image.py", line 34, in _get_elements
    return partition_image(filename=self.file_path, **self.unstructured_kwargs)
  File "/home/jon/miniconda3/envs/h2ollm/lib/python3.10/site-packages/unstructured/documents/elements.py", line 222, in wrapper
    elements = func(*args, **kwargs)
  File "/home/jon/miniconda3/envs/h2ollm/lib/python3.10/site-packages/unstructured/partition/image.py", line 52, in partition_image
    return partition_pdf_or_image(
  File "/home/jon/miniconda3/envs/h2ollm/lib/python3.10/site-packages/unstructured/partition/pdf.py", line 184, in partition_pdf_or_image
    layout_elements = _partition_pdf_or_image_local(
  File "/home/jon/miniconda3/envs/h2ollm/lib/python3.10/site-packages/unstructured/utils.py", line 43, in wrapper
    return func(*args, **kwargs)
  File "/home/jon/miniconda3/envs/h2ollm/lib/python3.10/site-packages/unstructured/partition/pdf.py", line 243, in _partition_pdf_or_image_local
    layout = process_file_with_model(
  File "/home/jon/miniconda3/envs/h2ollm/lib/python3.10/site-packages/unstructured_inference/inference/layout.py", line 370, in process_file_with_model
    model = get_model(model_name)
  File "/home/jon/miniconda3/envs/h2ollm/lib/python3.10/site-packages/unstructured_inference/models/base.py", line 41, in get_model
    model.initialize(**DETECTRON2_ONNX_MODEL_TYPES[model_name])
  File "/home/jon/miniconda3/envs/h2ollm/lib/python3.10/site-packages/unstructured_inference/models/detectron2onnx.py", line 75, in initialize
    self.model = onnxruntime.InferenceSession(
  File "/home/jon/miniconda3/envs/h2ollm/lib/python3.10/site-packages/onnxruntime/capi/onnxruntime_inference_collection.py", line 383, in __init__
    self._create_inference_session(providers, provider_options, disabled_optimizers)
  File "/home/jon/miniconda3/envs/h2ollm/lib/python3.10/site-packages/onnxruntime/capi/onnxruntime_inference_collection.py", line 435, in _create_inference_session
    sess.initialize_session(providers, provider_options, disabled_optimizers)
onnxruntime.capi.onnxruntime_pybind11_state.RuntimeException: [ONNXRuntimeError] : 6 : RUNTIME_EXCEPTION : Exception during initialization: /onnxruntime_src/onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.cc:1387 SubGraphCollection_t onnxruntime::TensorrtExecutionProvider::GetSupportedList(SubGraphCollection_t, int, int, const onnxruntime::GraphViewer&, bool*) const [ONNXRuntimeError] : 1 : FAIL : TensorRT input: /proposal_generator/GatherND_2_output_0 has no shape specified. Please run shape inference on the onnx model first. Details can be found in https://onnxruntime.ai/docs/execution-providers/TensorRT-ExecutionProvider.html#shape-inference-for-tensorrt-subgraphs
```

DocTR:

image

**Comment by pseudotensor** [comment]
> resolved issues related to this, but found another issue. Seems that the options selected dont propagate if you upload a zip. It always reverts to the default blip.

Fixed thanks.  Will add tests tomorrow, late now.

**Comment by ryanchesler** [comment]
updated with the new doctr version with onnx cleaned out

**Comment by pseudotensor** [comment]
All new ~100 tests pass except weaviate hit timeout and I'll ignore that.

image

**Review by pseudotensor — APPROVED** [review]

**Pull request merged**